### 并发控制

> [!同时最多执行n个函数，控制执行数量]

``` javascript
class pLimit {  
    maxCount = 5  
    currentCount = 0  
    taskList = []  
  
    constructor(maxCount) {  
        this.maxCount = maxCount  
        this.run = this.run.bind(this)  
    }  
    addTask(fn, ...args) {  
        this.taskList.push(() => fn(...args))  
    }  
    run() {  
        while (this.currentCount < this.maxCount && this.taskList.length > 0) {  
            const fn = this.taskList.shift()  
            if (fn) {  
                this.currentCount++  
                const result = fn()  
                result.finally(() => {  
                    this.currentCount--  
                    this.run()  
                })  
            }  
        }  
    }  
}
```
##### 使用示例
```javascript 
const limit = new pLimit(3)  
  
function createTask(time) {  
    return new Promise((resolve) => {  
        setTimeout(() => {  
            console.log(`Task with ${time}ms delay done`)  
            resolve()  
        }, time)  
    })  
}  
  
limit.addTask(createTask, 1000)  
limit.addTask(createTask, 1000)  
limit.addTask(createTask, 1000)  
limit.addTask(createTask, 1000)  
limit.addTask(createTask, 2000)  
limit.addTask(createTask, 1000)  
limit.addTask(createTask, 1000)  
limit.addTask(createTask, 1000)  
limit.addTask(createTask, 3000)  
limit.addTask(createTask, 1000)  
limit.run()
```

### 请求节流

```javascript
const requestThrottle = function (promiseFun, key, options = {}) {  
    const {  
        // 节流期限时间  
        timestamp = 3000,  
        // 是否强制重新请求  
        forceUpdate = false,  
        onError = null,  
        cache = requestThrottle.pool || (requestThrottle.pool = {}) // 缓存池  
    } = options  
  
    if (forceUpdate) delete cache[key]  
  
    if (!cache[key]) {  
        try {  
            const result = promiseFun()  
            cache[key] = result  
            result  
                .then((response) => {  
                    return response // 将结果传递给下一个then 调用函数可以通过then直接拿到这个结果  
                })  
                .catch((err) => {  
                    delete cache[key]  
                    if (onError) onError(err)  
                    throw err // 保证发生错误的时候，调用函数仍然能catch错误信息  
                })  
  
            // 防抖处理  
            const _key = `${key}_timeout`  
            if (cache[_key]) clearTimeout(cache[_key])  
            cache[_key] = setTimeout(() => delete cache[key], timestamp)  
        } catch (err) {  
            if (onError) onError(err)  
            throw err // 保证发生错误的时候，调用函数仍然能catch错误信息  
        }  
    } else {  
        return cache[key]  
    }  
}
```
